react/features/chat/middleware.js

```diff
@@ -116,44 +152,183 @@ function _addChatMsgListener(conference, { dispatch, getState }) {
-                    body: message,
-                    id,
-                    nick: displayName,
-                    ts: timestamp
-                });
+    // Provide a default for for the case when a message is being
+    // backfilled for a participant that has left the conference.
+    const participant = getParticipantById(state, id) || {};
+    const localParticipant = getLocalParticipant(getState);
+    const displayName = participant.name || nick || getParticipantDisplayName(state, id);
+    const hasRead = participant.local || isChatOpen;
+    const timestampToDate = timestamp
+        ? new Date(timestamp) : new Date();
+    const millisecondsTimestamp = timestampToDate.getTime();
 
-                dispatch(showToolbox(4000));
-            }
-        }
-    );
+    dispatch(addMessage({
+        displayName,
+        hasRead,
+        id,
+        messageType: participant.local ? 'local' : 'remote',
+        message,
+        privateMessage,
+        recipient: getParticipantDisplayName(state, localParticipant.id),
+        timestamp: millisecondsTimestamp
+    }));
+
+    if (typeof APP !== 'undefined') {
+        // Logic for web only:
+
+        APP.API.notifyReceivedChatMessage({
+            body: message,
+            id,
+            nick: displayName,
+            ts: timestamp
+        });
+
+        dispatch(showToolbox(4000));
+    }
+}
+
+/**
+ * Focuses the chat text field on web after the message recipient was updated, if needed.
+ *
+ * @returns {void}
+ */
+function _maybeFocusField() {
+    if (navigator.product !== 'ReactNative') {
+        const textField = document.getElementById('usermsg');
+
+        textField && textField.focus();
+    }
+}
+
+/**
+ * Persists the sent private messages as if they were received over the muc.
+ *
+ * This is required as we rely on the fact that we receive all messages from the muc that we send
+ * (as they are sent to everybody), but we don't receive the private messages we send to another participant.
+ * But those messages should be in the store as well, otherwise they don't appear in the chat window.
+ *
+ * @param {Store} store - The Redux store.
+ * @param {string} recipientId - The recipient the private message was sent to.
+ * @param {string} message - The sent message.
+ * @returns {void}
+ */
+function _persistSentPrivateMessage({ dispatch, getState }, recipientId, message) {
+    const localParticipant = getLocalParticipant(getState);
+    const displayName = getParticipantDisplayName(getState, localParticipant.id);
+
+    dispatch(addMessage({
+        displayName,
+        hasRead: true,
+        id: localParticipant.id,
+        messageType: 'local',
+        message,
+        privateMessage: true,
+        recipient: getParticipantDisplayName(getState, recipientId),
+        timestamp: Date.now()
+    }));
+}
+
+/**
+ * Returns the ID of the participant who we may have wanted to send the message
+ * that we're about to send.
```

Comment:
Is this function shouldShowPrivacyMessage?

Reply:
Not really, because the code decides there, if there is a participant that we may have intended to send a private message to. Then That is the basis to decide to show the notice for sure, but that's another step. Or at least that is my thinking of it